### DISCUSSION TITLE
Is it possible to force casting to be applied

### DISCUSSION THREAD

**Discussion by splatEric** [spatie/laravel-data#352] [Q&A] [answered]
Hi,

I'm in the process of trying to move from `spatie/data-transfer-object` to this package, and have hit a small challenge with casting.

My DTO looks like this:

```
class PatientSearchResult extends Data
{
    use HasFactory;

    public function __construct(

        #[WithCast(PatientCollectionCaster::class)]
        public PatientCollection $patients,
        public string $sourceId,
        public bool $success,

        #[WithCast(CarbonImmutableCaster::class)]
        public CarbonImmutable $retrievedAt
    ) {
    }
}
```

with the `CarbonImmutableCaster`:

```
class CarbonImmutableCaster implements Cast
{
    public function cast(DataProperty $property, mixed $value): mixed
    {
        if ($value instanceof CarbonImmutable) {
            return $value;
        }

        if ($value instanceof DateTime) {
            return CarbonImmutable::createFromMutable($value);
        }

        if (is_string($value)) {
            return CarbonImmutable::parse($value);
        }

        return new Uncastable();
    }
}
```

if I try to do the following:

```
PatientSearchResult::from([
    // ...
   'retrievedAt' => new DateTime()
]);
```

I get the following exception:

` TypeError  App\Services\PatientSearch\DTO\PatientSearchResult::__construct(): Argument #4 ($retrievedAt) must be of type Carbon\CarbonImmutable, DateTime given, called in vendor/spatie/laravel-data/src/Resolvers/DataFromArrayResolver.php on line 90.`

If I pass a string in, then the casting works fine. Am I misunderstanding how I should cast a property from other class types?

**Comment by rubenvanassche** [comment, marked as answer]
Could you try this again with the newest version of the package?

I cannot reproduce it:

```php
it('cacas', function () {
    class CarbonImmutableCaster implements Cast
    {
        public function cast(DataProperty $property, mixed $value, array $context): mixed
        {
            if ($value instanceof CarbonImmutable) {
                return $value;
            }

            if ($value instanceof DateTime) {
                return CarbonImmutable::createFromMutable($value);
            }

            if (is_string($value)) {
                return CarbonImmutable::parse($value);
            }

            return new Uncastable();
        }
    }

    $dataClass = new class extends Data {
        #[WithCast(CarbonImmutableCaster::class)]
        public CarbonImmutable $retrievedAt;
    };

    dd($dataClass::from([
        'retrievedAt' => new DateTime()
    ]));
});
```

Seems like you're still using v1 if I'm correct? A lot has changed since then 😄

**Reply by splatEric** [reply]
I had not realised I was still on version 1.x ... it's because I have been a lazy developer and not yet upgraded to php 8.1! I will deal with that first, and then come back to this problem. In the meantime, thank you very much for coming back to me, much appreciated.

**Reply by splatEric** [reply]
yep, upgraded, and it's now working. Sorry for time wasting.
